content/blog/2019-05-14-nasapower.md

```diff
@@ -0,0 +1,625 @@
+
+I quickly found out that the easy part was using crul; the hard part
+was validating user inputs. I decided early in the process of writing
+the package to validate the users’ input on the client side before even
+querying the server to make sure that only well-formed requests were
+sent. That way I could provide the user with feedback on what may have
+been entered incorrectly to make it easier for them to correct rather
+than relying on the server’s response.
+
+There are over 140 different parameters that the POWER data provides.
+One of the first issues I encountered was how to validate the users’
+requests against what was available. Thankfully [Ben
+Raymond](https://github.com/raymondben) found a JSON file that was
+served up for the API documentation that I was able to use to [create an
+internal list
+object](https://github.com/ropensci/nasapower/tree/master/data-raw) to
+check against, which made the parameters easy to check against.
+Formatting dates and checking those given all the different ways we
+enter them proved to be another challenge entirely taking nearly 100
+lines of code.
+
+Along with learning how to write an API client package, one of the
+methods I used in this package that I had not made full use of before
+was lists. Internally in that 100+ lines of code there are several
+inter-related checks and values that are provided for the query. By
+using lists I was able to return more than one value from a function
+after it was checked and validated and then provide that to the function
+I created to generate the request that crul sends, *e.g.*, the
+[longitude, latitude and whether it is a single point or
+region](https://github.com/ropensci/nasapower/blob/992c99d45f3a42471664e67db8921caf74dbc90c/R/internal_functions.R#L237).
+By doing this I was able to simplify the parameters that the user had to
+enter when constructing a query in their R session because the API
+dictates by default many values that cannot co-occur.
+
+### Are We Doing it Right?
+
+There were a few hang-ups along the way. At one point the POWER team
+made a change to the API response while the package was in review.
+Initially it provided an option for just a CSV file that was properly
+rectangular. Suddenly it changed overnight with no warning to offering a
+CSV file with information in a “header” as they called it. To me a
+header in a CSV file is just column names. This was more. This was
+metadata about the data itself. The quickest way to deal with this was
+to simply read only the CSV file portion of the returned file. However,
+I thought it might be useful to users to include the metadata that POWER
+was now providing in this format as well. So I learned something else
+new, how to modify the S3 `print()` method function to include the
+metadata in the console along with the weather or climate data.
+
+Later, in early 2019 I ran into an issue with the method I’d used to
+validate user inputs for `community` when building the query that was
```

Comment:
It wasn't really clear to me what `community` refers to, e.g. communities like "agricultural weather data (AG community)" in first section.  Maybe it wasn't clear to me that there would be multiple "communities" to do user input check. Please clarify

Reply:
I've attempted to clarify this. But won't resolve it yet until we're good. :)